src/emqtt_ws.erl

```diff
@@ -40,9 +40,11 @@ connect(Host0, Port, Opts, Timeout) ->
     Host1 = convert_host(Host0),
     {ok, _} = application:ensure_all_started(gun),
     %% 1. open connection
+    TcpOpts = tcp_opts(Opts, []),
     TransportOptions = proplists:get_value(ws_transport_options, Opts, []),
```

Comment:
Yeah, I think it's currently (perhaps accidentally) designed to be completely separate: `gen_tcp`-level options are supplied through `tcp_opts` when it's over TCP, but when it's over WS the options are in `ws_transport_options`.

I guess if we want to unify that, we also need to take care of `ssl` and `ssl_opts` translation. And ideally disentangle `#state.sock_opts` on `emqtt` module level to avoid dealing with this weird option whilelisting below.

Reply:
Note though, that one cannot add, for example, `{ip, ...}`, to `ws_transport_options`, since all these options are added flat to `ConnOpts`. The checks in `gun` will fail.

Reply:
I mean, in this case it'll look like `{ws_transport_options, [{tcp_opts, [{ip, ...}]}]}`, see for example https://github.com/emqx/emqtt/blob/54d027d6df0073201e023d7b68b97809a9549f4c/test/emqtt_connect_SUITE.erl#L37-L38.

Reply:
I'll close this in favor of proper patch.